### ISSUE TITLE
Cannot resolve a reference with multiple candidates (two model instances with same identifier)

### ISSUE THREAD

**Issue by zisiszikos** [mobxjs/mobx-state-tree#1156]
**_Bug report_**

* [x] I've checked documentation and searched for existing issues
* [x] I've made sure my project is based on the latest MST version
* [ ] Fork [this](https://codesandbox.io/s/y64pzxj01) code sandbox or another minimal reproduction. 

**Sandbox link or minimal reproduction code**
[Sandbox example](https://codesandbox.io/s/2x10pl6j4n)

**Describe the expected behavior**
I have a `Base` model that is composed in two places `one, two` in the store tree. This model uses nested models `Country` and `City`. I add a new country and a new city in both places with the same identifiers. The cities are also placed in an array of references at the country model. Now I want to get the cities of a country by reading the list of cities references.

**Describe the observed behavior**
When trying to get the list of cities from references list of `Country` model, the following error throws:
```
[mobx-state-tree] Cannot resolve a reference to type 'late(function () {
    return City;
  })' with id: '1000' unambigously, there are multiple candidates: /one/cities/1000, /two/cities/1000
```
The same is NOT happening when trying to get a country, even though there are multiple countries with the same identifier, but not in a list of references.

It seams that resolving an item by its reference identifier not working if there are multiple items with the same identifier.

**Comment by xaviergonz** [comment]
https://github.com/mobxjs/mobx-state-tree/blob/master/README.md#references

Identifiers for objects of the same kind must be unique across the whole tree that spawns the whole root store, this is because in their snapshot form references are saved as plain strings /numbers that must point to a single object. 

There are ways to escape this such as custom references or views that resolve the nodes in a custom way.

Not sure if that's what you meant though 🙂

**Comment by zisiszikos** [comment]
Thanks for the answer.

I understand what you are saying but there might be situations where a model needs to be used in multiple places in the store. For example, imagine a `User` model that can be used for regular product users and for administrator users, both stored at different arrays. The only solution now is to duplicate the model.

Anyway, at the time, I solved my problem by replacing the `safeReference` type property with a view getter that returns the same thing.

[New example here](https://codesandbox.io/s/olvvl34z2y)

**Comment by xaviergonz** [comment]
Another way to tackle the problme would be to use a custom reference that resolves to a certain array in the store rather than relying on the built-in resolveIdentifier.

To be honest I don't use references that much, I usually just use ids as properties and views to resolve them :)

**Comment by xaviergonz** [comment]
Btw, if you have any suggestion on how MST should tackle the problem instead of the current implementation feel free to do so!

**Comment by kingpalethe** [comment]
This is a useful thread. What I am confused about is this:  What is MST's recommended workflow for **deep cloing** a tree that uses `reference` or `safeReference`?

In my case I have a complex structure where a parent references a child and that child references another child, and each contains references to objects below in the tree....... It would seem I would either need to:

1) give up on `reference` and `safeReference`, like @zisiszikos did above
OR
2) When-ever I clone part of a tree, I need to take a snapshot of the part of tree I want to clone, and iterate over all the children in this tree, setting up "new" children each time, copying all the data into them but creating new unique ids for each one?

**Comment by rseyferth** [comment]
I agree, this is an important issue. Having the application break when you try to use a model in 2 locations is really unhelpful... Especially because there is no way to add a 'generic' model property to a model, so you can abstract some logic, and `A node cannot exists twice in the state tree`. It's not an uncommon thing for a record to appear in an application twice, right?

What would be the recommended way to reference generic models? The use case I have is that a user can select elements from different pages to add them to a map, so I'd need an array of references to models, and not know exactly how they fit in the tree... Is there a way to do this?

**Comment by filipkis** [comment]
I'm having a similar issue (same User appearing in various arrays) except I'm not using references, but just `resolveIdentifier`. However, I just need to resolve it to get the user name and I don't care which instance (i.e. from which array) I get it.

And I was wondering, would it be possible to include a `resolveIdentifierAll` (or some other name) function or add an optional parameter that wouldn't throw an error if there are multiple instances with same id. For my case it would be fine if the function worked as `find`, i.e. returning the first.

**Comment by mweststrate** [comment]
> It's not an uncommon thing for a record to appear in an application twice, right?

It is; if an element appears twice, it means that you didn't create a _tree_ which all other utilities assume (not having a tree for examples makes it impossible to serialize reliably, or to determine unambigously who is the parent of a node). If you find yourself needing to have the same node multiple times in the tree, probably all instances of that except one should be references.

For further questions, please open a new issue, as this one is closed already and others won't see the question.

**Comment by jamonholmgren** [comment]
I wouldn't mind a more elegant solution to this issue, as it's something we've run into ourselves. Re-opening for future consideration.

**Comment by 10000multiplier** [comment]
@jamonholmgren were you able to find a solution?

**Comment by Navipro70** [comment]
Maybe use types.string instead of types.identifier helps?

**Comment by 10000multiplier** [comment]
@Navipro70 will that still work with references?

**Comment by Navipro70** [comment]
@10000multiplier 
I don't know, try to test it)

**Comment by filipkis** [comment]
@10000multiplier and @Navipro70 no it won't that's the whole point of identifier.

**Comment by jamonholmgren** [comment]
I've released a new library to help with this problem, which I think is pretty cool, well-tested, and pretty well documented.

Please check it out and let me know what you think about it!

It's called mst-reference-pool and you can find it here:

https://github.com/infinitered/mst-reference-pool

In a nutshell, it gives you tools to easily create a pool of instances, and then you simply use references from anywhere on the tree. It also comes with a garbage collector so you don't end up with a memory leak when instances are no longer referenced from anywhere.

**Comment by andrewdang17** [comment]
@jamonholmgren that looks great, will definitely give it a try!

I am curious though how this would be handled if a tool like that didn't exist. Seems like MST has been out for a long time that this would be accounted for and documented? My only thought is to not use references at all like some others suggested but then what's the point of even using MST.

**Comment by coolsoftwaretyler** [comment]
Hey @jamonholmgren, et. al., looks like `mst-reference-pool` is a good add-on fix here. I'm going to convert this issue to a discussion because I think there's some opportunity for us to consider how that approach might make sense for MST (I also need this in my own application), but the "issue" seems to be resolved for sure.

Thanks to everyone who chimed in!
